### ISSUE TITLE
Full world simulation issues

### ISSUE THREAD

**Issue by Try** [Try/OpenGothic#657]
#### Why full world simulation:
* Goes very well along limitless view distance, resulting in better game experience
* Gives us a reason to say that npc runtime doesn't degradate over time (good for testing)
* and just cool overall

#### Known issues
* Earthquake (this issue)
  * Essentially any other global call from script, akin to `WLD_PLAYEFFECT` will cause trouble
* https://github.com/Try/OpenGothic/issues/585
  * Ring of water meeting(Lares and others travel in real-time)
  * Greg doesn't come quick enough to kill Dexter
  * Special-casing `ai_gotowp("TOT")`, due to Pedro in gothic2
* https://github.com/Try/OpenGothic/issues/354
  * NPCs won't heal in time
* Npc state doesn't refresh(part of https://github.com/Try/OpenGothic/pull/653)
* `Npc::tickRoutine` has no call to LOOP function for far-away monsters (performance)

#### Proposed solutions

1. Suppress `_LOOP` function for `AiFar`/`AiFar2`.
Pros: very close to original game, where far npc's are offloaded
Cons: wont be able to see routine animations at a distance, such as small-talk and practice_sword.
Cons: if mod has side-effect in non-loop part if AI - wont work as intended 

2. Suppress 'side-effects' of _LOOP function
Pros: solves immediate problem of earthquakes
Cons: hard to implement (need to do deep tracking of context, or check call-stack for `*_LOOP`)
Cons: need to solve cases 1 by one, and it's possible (at least on purpose) to break this solution with any level of indirection

**Comment by Abendlied** [comment]
You are refering to this as a solution for e.g. the Greg to Dexter example, so I assume _LOOP influences moving NPCs after story triggers. Are escorting activities also impacted by the proposals? If so, a lof things are potentially affected by this change, so I propose to make a list that aims to help understand whether we cover all relevant cases with the proposed solution.
I am not aware of anything G2 classic does that is not included in G2 NotR, so I left G2 classic it out.
Also, I cannot really well differentiate in G2 which is a spawn/"despawn" (aka killed and stored in caves) of a new NPC version and what is movement of same entity.

**Escorting NPCs**
*Gothic 1:*
[ ] Lares to new camp
[ ] Baal Parvez to sect camp
[ ] Talas to goblin cave and to orc cementry

*Gothic 2 NotR:*
[ ] Lares to Onar (with intermittent waypoints)
[ ] Lares to stone ring (with intermittent waypoints)
[ ] Lares to water mages (with intermittent waypoints)
[ ] Mika to Akil
[ ] Bartok hunting (with intermittent waypoints)
[ ] Alligator Jack to pirate camp

**NPCs moving position after story trigger**
*Gothic 1:*
[ ] Diego from start lake to old camp (kills all starter monsters on the way)
[ ] Diego, Milten, Gorn, Lester for focus quest
[ ] Gorn for new mine recapturing
[ ] Lee to dam for monologue
[ ] Diego, Lester, Scorpio after chapter 4 triggered and OC gates are closed
[ ] Milten to NC watermages
[ ] Ur-Shak to orc city
[ ] Bartholo and Raven after Teleport to OC

*Gothic 2 NotR:*
_Khorinis_
[ ] Canthar to city
[ ] Greg to city
[ ] Greg to Orlans
[ ] Greg to Onar road
[ ] Greg to Dexter
[ ] Erol to his hut
[ ] Attila to hut behind Halvor
[ ] Ring of water meeting
[ ] Diego, Lothar into city
[ ] Xardas, Pyrokar, Vatras Eye of Innos at stonehenge
[ ] Lester to Xardas
[ ] Jharkendar prisoners to Khorinis
[ ] Rengaru running
[ ] Jack to lighthouse
[ ] Brian to lighthouse
[ ] Chosen crew to harbour after reading sea map 
[ ] Chosed NPCs to prison (Canthar, Sarah, Rengaru, Fernando...) - and out again
[ ] Salandril to monestary
[ ] Bennet to prison and back to mercenaries
[ ] Elvrich nephew apprentice back to Thorben
[ ] Randolph for the drinking quest
[ ] Cavalorn to the city
[ ] Gerbrandt to harbour district

_Jharkendar_
[ ] Fisk and bodyguards to Snafs
[ ] Bloodwyn to minecrawler cave

_VoM_
[ ] Angar 
[ ] multiple NPCs to old camp
[ ] Ur-Shak to Hosh-Pak place
probably much more I cannot think of right now

_Irdorath_
[ ] Pedro 
[ ] Mario

**Comment by Abendlied** [comment]
Another point would be aggro distance for NPC vs. monster.

In Gothic 1, attacking a swamp shark, the monster will come after me for a long distance until sect camp inhabitants kill it.

In Gothic 2 NotR, transforming to dragon snapper in harbour district and swimming to treasure island, Khorinis inhabitants will come after me even after minutes, swimming all the way to me.

**Comment by Try** [comment]
> You are refering to this as a solution for e.g. the Greg to Dexter example, so I assume _LOOP influences moving NPCs after story triggers. 

No, maybe I was unclear. `_LOOP` based solutions will prevent unintended side-effects, such as earthquake but wont affer walk-teleport problem (walk-teleport is another issues here: https://github.com/Try/OpenGothic/issues/585)

> Another point would be aggro distance for NPC vs. monster.

In opengothic aggro/notify distance is exactly same as in vanilla. Chaise time/distance might diverge in cases, when monster doesn't have any condition in a script to stop chaise.

Example can be `ZS_MM_Hunt_Loop`: monsters, such as snapper will chaise pray until it offloaded from memory. In Game of opengothic - it's means forever chaise.

> Third point: roaming monsters fall off cliffs.

Should be fixed nowdays by changing on how cliff-collision works, but need to test

**Comment by Abendlied** [comment]
> In opengothic aggro/notify distance is exactly same as in vanilla. Chaise time/distance might diverge in cases, when monster doesn't have any condition in a script to stop chaise.
> 
> Example can be `ZS_MM_Hunt_Loop`: monsters, such as snapper will chaise pray until it offloaded from memory. In Game of opengothic - it's means forever chaise.

With your explanation, I can confirm I mean indeed chasing distance, not aggro distance.




> Example can be `ZS_MM_Hunt_Loop`: monsters, such as snapper will chaise pray until it offloaded from memory. In Game of opengothic - it's means forever chaise.

I observed snappers at black troll using `ztimer multiplier X`. They do roam around and seem to chase, passing over the brindge and beyond until out of sight.



> Should be fixed nowdays by changing on how cliff-collision works, but need to test

Was actually not an issue for the snappers, seems cliff-collision works better now.

**Comment by Abendlied** [comment]
I need to correct myself in both observations.

> I observed snappers at black troll using `ztimer multiplier X`. They do roam around and seem to chase, passing over the brindge and beyond until out of sight.

Not chasing in this case. I suspect it is their animation loop that moves them forward a tiny bit, after a while it accumulates. Directions never change. Either that or some side effect of `ztimer multiplier X`.

video coast, 13 sec, 8.1 MiB: https://github.com/user-attachments/assets/fb1df3c0-4c8f-4a97-8183-17508a4ebc90--
video at black troll, 6 sec, 6.3 MiB: https://github.com/user-attachments/assets/30a86740-645a-440f-8354-f34f2ddce5ae

> > 
> 
> Was actually not an issue for the snappers, seems cliff-collision works better now.

they do stop at the cliff. However, normal animation movement causes them to fall still.




video coast snapper falling, 13 sec, 8.1 MiB: ![Falling Snapper](https://github.com/user-attachments/assets/61f9146c-6e04-465b-9a6a-772ce5e1b7d3)

**Comment by Try** [comment]
For snapper persistent offset from origin seem to be related to animation + `ZS_MM_RTN_REST`.
In the script:
```c++

func int ZS_MM_Rtn_Rest_Loop()
{
	...
	if(self.aivar[AIV_TAPOSITION] == NOTINPOS)
	{
		if(Wld_IsFPAvailable(self,"FP_ROAM"))
		{
			AI_GotoFP(self,"FP_ROAM");
		};
		if(Npc_IsOnFP(self,"FP_ROAM"))
		{
			self.aivar[AIV_TAPOSITION] = ISINPOS;
		};
	}
	...
```

Since OpenGothic skips `_LOOP`, for far away monsters no `AI_GotoFP` corrections was performed, while animation continued to accumulate offset.

**Comment by 10JML01** [comment]
Does the Snapper moving bug still exist in 1.0.3010?

**Comment by Try** [comment]
Hi, @10JML01 !

At the moment there is no good ideas on how correct behavior can be like: send rare ticks? halt animation? else?
So no proper fix; only small cancellation: snapper wont fall from the cliff.

**Comment by 10JML01** [comment]
Thanks for the swift reply. So Snappers should not fall off any cliff any more in the whole game? I don't remember reading about the "small cancellation", what did you change?
Would it be possible to adjust the animation to have the step forward removed? Does this only affect Snappers, or are e.g. Orcs also affected by having a forward step in their animation?

**Comment by Try** [comment]
> Would it be possible to adjust the animation

Not really: engine only plays animation given to it, and can only ignore offset. But ignoring offset has large impact on overall game, in normal cases.

> Does this only affect Snappers

Any npc technically. However, usually humans with daily time-schedule have no opportunity to run into a problem.

> "small cancellation", what did you change?

There where some changes in physics in the mean time, so npc got stuck on the edge more reliably (similar to walk mode for player)

**Comment by 10JML01** [comment]
I wonder if Grimbald will still be pissed off at me because I didn't help him with any snappers (in case them chilling at the edge of the cliff is too far away for him to count them).
Will report when I've been there.
Edit 2nd Oct 2025: Grimbald not impressed. One Snapper at mountain wall (which he promptly shot dead) and two snappers found below the bridge in the water. Looks like the fix did not work properly. @Try

**Comment by Try** [comment]
Hi @10JML01 !

Here is [c38bd65](https://github.com/Try/OpenGothic/commit/c38bd65830addbf2736fdb4ddc60ea289c910bcb) newer approach to npc scheduler. Quite a bit more expensive(performance), that we have had before, yet seem to work good enough.

While CI id down, would be hard to shared any build, unfortunately. Are you capable of building from source?

**Comment by Try** [comment]
Based on recent release `_LOOP` based approach doesn't work in many cases. Script doesn't have strong enough correction on restart:

```
func void ZS_MM_Rtn_Roam()
{
	...
	if (Hlp_StrCmp(Npc_GetNearestWP(self),self.wp)==FALSE)  //damit die Monster beim Inserten nicht immer erst zum WP rennen, sondern nur, wenn sie der Heimat zu fern sind
	{
		AI_GotoWP (self, self.WP);
	};	
	...
};
```

On cliff edge closes point may not change, until monster fall down.

**Comment by Try** [comment]
New heuristic:
* track if script was invoked due to AI-processing: to filter-out Saturas case
* check distance between npc and it's way-point. If more than 30meters (arbitrary) - sent a tick to AI.

3 meters seem to be good compromise - not too far to fall-apart, not too close to be a performance problem.
